app/auth/oauth_client.py

```diff
@@ -0,0 +1,129 @@
+# limitations under the License.
+
+import json
+import sys
+import time
+
+from oauthlib.oauth2 import WebApplicationClient
+from oauthlib.oauth2.rfc6749.errors import OAuth2Error
+from requests_oauthlib import OAuth2Session
+
+from .oauth_provider_app import OAuthProviderApp, JupyterHubProviderApp
+
+
+class RenkuWebApplicationClient(WebApplicationClient):
+    """WebApplicationClientClass enriched with provider/app information and
+    methods for (de-)serializing and obtaining tokens from the provider."""
+
+    def __init__(
+        self,
+        *args,
+        provider_app=None,
+        scopes=[],
+        max_lifetime=None,
+        _expires_at=None,
+        **kwargs
+    ):
+        super().__init__(provider_app.client_id, *args, **kwargs)
+        assert isinstance(
+            provider_app, OAuthProviderApp
+        ), "provider_app property must be instance of OAuthProviderApp class"
+        self.provider_app = provider_app
+        self.scopes = scopes
+        self.max_lifetime = max_lifetime
+        self._expires_at = _expires_at
+
+    def prepare_authorization_request(self):
+        """Overwrite parent method to work with provider properties
+        instead of an argument"""
+        return super().prepare_authorization_request(
+            self.provider_app.authorization_endpoint
+        )
+
+    def fetch_token(self, authorization_response, **kwargs):
+        """Convenience method for fetching tokens."""
+        oauth_session = OAuth2Session(client=self, redirect_uri=self.redirect_url)
+        oauth_session.fetch_token(
+            self.provider_app.token_endpoint,
+            authorization_response=authorization_response,
+            client_secret=self.provider_app.client_secret,
+            client_id=self.provider_app.client_id,
+            scopes=self.scopes,
+            include_client_id=True,
+            **kwargs
+        )
+        # TODO: Once Jupyterhub actually expires access tokens and implements
+        # TODO: usable refresh tokens, this subclass can be removed.
+        # Could use a factory method instead, but seems like an overkill...
+        if isinstance(self.provider_app, JupyterHubProviderApp):
+            self.refresh_token = None
+            self.expires_in = None
+        self._fix_expiration_time()
+
+    def refresh_access_token(self):
+        """Convenience method for refreshing tokens."""
+        self._expires_at = None
+        oauth_session = OAuth2Session(client=self)
+        oauth_session.refresh_token(
+            self.provider_app.token_endpoint,
+            client_id=self.provider_app.client_id,
+            client_secret=self.provider_app.client_secret,
+            include_client_id=True,
+        )
+        self._fix_expiration_time()
+
+    # Note: Pickling would be much simpler here, but we don't fully control
+    # what's going to be pickeled, so we choose the safer approach.
+    def to_json(self):
```

Comment:
This could be nicely done with marshmallow. Potential improvement / TODO ticket?

Reply:
https://github.com/SwissDataScienceCenter/renku-gateway/issues/231